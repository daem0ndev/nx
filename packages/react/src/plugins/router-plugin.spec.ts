import { CreateNodesContext } from '@nx/devkit';
import { createNodes } from './router-plugin';
import { TempFs } from 'nx/src/internal-testing-utils/temp-fs';
import { isUsingTsSolutionSetup } from '@nx/js/src/utils/typescript/ts-solution-setup';
import { loadConfigFile } from '@nx/devkit/src/utils/config-utils';

jest.mock('../utils/load-vite-dynamic', () => ({
  loadViteDynamicImport: jest.fn().mockResolvedValue({
    resolveConfig: jest.fn().mockResolvedValue({}),
  }),
}));

jest.mock('@nx/devkit/src/utils/config-utils', () => {
  return {
    ...jest.requireActual('@nx/devkit/src/utils/config-utils'),
    loadConfigFile: jest.fn(),
  };
});

jest.mock('@nx/js/src/utils/typescript/ts-solution-setup', () => ({
  ...jest.requireActual('@nx/js/src/utils/typescript/ts-solution-setup'),
  isUsingTsSolutionSetup: jest.fn(),
}));

describe('@nx/react/react-router-plugin', () => {
  let createNodesFunction = createNodes[1];
  let context: CreateNodesContext;
  let cwd = process.cwd();

  beforeEach(() => {
    (isUsingTsSolutionSetup as jest.Mock).mockReturnValue(false);
  });

  describe('React Router', () => {
    describe('root project', () => {
      const tempFs = new TempFs('temp');

      beforeEach(() => {
        context = {
          nxJsonConfiguration: {
            namedInputs: {
              default: ['{projectRoot}/**/*'],
              production: ['!{projectRoot}/**/*.spec.ts'],
            },
          },
          workspaceRoot: tempFs.tempDir,
          configFiles: [],
        };

        tempFs.createFileSync(
          'package.json',
          JSON.stringify({ name: 'acme', type: 'module' })
        );
        tempFs.createFileSync('react-router.config.js', JSON.stringify({}));
        tempFs.createFileSync('vite.config.js', JSON.stringify({}));
      });

      afterEach(() => {
        jest.resetModules();
      });

      it('should create nodes by default', async () => {
        const configPath = 'react-router.config.js';
        (loadConfigFile as jest.Mock).mockReturnValue({});
        mockConfig('vite.config.js', {});

        const nodes = await createNodesFunction(
          [configPath],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
          },
          context
        );

        expect(nodes).toMatchSnapshot();
      });

      it('should create nodes without start target if ssr is false', async () => {
        const configPath = 'react-router.config.js';
        (loadConfigFile as jest.Mock).mockReturnValue({ ssr: false });
        mockConfig('vite.config.js', {});

        const nodes = await createNodesFunction(
          [configPath],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
          },
          context
        );

        expect(nodes).toMatchSnapshot();
      });
    });

    describe('integrated projects', () => {
      let tempFs: TempFs;

      beforeEach(() => {
        tempFs = new TempFs('test');
        context = {
          nxJsonConfiguration: {
            namedInputs: {
              default: ['{projectRoot}/**/*'],
              production: ['!{projectRoot}/**/*.spec.ts'],
            },
          },
          workspaceRoot: tempFs.tempDir,
          configFiles: [],
        };

        tempFs.createFileSync(
          'acme/project.json',
          JSON.stringify({ name: 'acme' })
        );

        tempFs.createFileSync('acme/react-router.config.js', '');
        tempFs.createFileSync('acme/vite.config.js', '');
      });

      afterEach(() => {
        jest.resetModules();
        tempFs.cleanup();
      });

      it('should create nodes by default', async () => {
        (loadConfigFile as jest.Mock).mockReturnValue({});
        mockConfig('acme/vite.config.js', {});

        const nodes = await createNodesFunction(
          ['acme/react-router.config.js'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
          },
          context
        );

        expect(nodes).toMatchSnapshot();
      });

      it('should create nodes without start target if ssr is false', async () => {
        (loadConfigFile as jest.Mock).mockReturnValue({ ssr: false });
        mockConfig('acme/vite.config.js', {});

        const nodes = await createNodesFunction(
          ['acme/react-router.config.js'],
          {
            buildTargetName: 'build',
            devTargetName: 'dev',
            startTargetName: 'start',
          },
          context
        );

        expect(nodes).toMatchSnapshot();
      });
    });
  });
});

function mockConfig(path: string, config: any) {
  jest.mock(
    path,
    () => ({
      default: config,
    }),
    {
      virtual: true,
    }
  );
}
