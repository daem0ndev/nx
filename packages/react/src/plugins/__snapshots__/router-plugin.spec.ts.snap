// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nx/react/react-router-plugin React Router integrated projects should create nodes by default 1`] = `
[
  [
    "acme/react-router.config.js",
    {
      "projects": {
        "acme": {
          "metadata": {},
          "root": "acme",
          "targets": {
            "build": {
              "cache": true,
              "command": "react-router build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "react-router",
                  ],
                },
              ],
              "options": {
                "cwd": "acme",
              },
              "outputs": [
                "{workspaceRoot}/acme/build/client",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "react-router dev",
              "options": {
                "cwd": "acme",
              },
            },
            "start": {
              "command": "react-router-server build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "acme",
              },
            },
            "typecheck": {
              "cache": true,
              "command": "tsc --noEmit",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "metadata": {
                "description": "Runs type-checking for the project.",
                "help": {
                  "command": "pnpm exec tsc --help",
                  "example": {
                    "options": {
                      "noEmit": true,
                    },
                  },
                },
                "technologies": [
                  "typescript",
                ],
              },
              "options": {
                "cwd": "acme",
              },
            },
            "watch-deps": {
              "command": "pnpm exec nx watch --projects acme --includeDependentProjects -- pnpm exec nx build-deps acme",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/react/react-router-plugin React Router integrated projects should create nodes without start target if ssr is false 1`] = `
[
  [
    "acme/react-router.config.js",
    {
      "projects": {
        "acme": {
          "metadata": {},
          "root": "acme",
          "targets": {
            "build": {
              "cache": true,
              "command": "react-router build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "react-router",
                  ],
                },
              ],
              "options": {
                "cwd": "acme",
              },
              "outputs": [
                "{workspaceRoot}/acme/build/client",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "react-router dev",
              "options": {
                "cwd": "acme",
              },
            },
            "start": {
              "command": "react-router-server build/server/index.js",
              "dependsOn": [
                "build",
              ],
              "options": {
                "cwd": "acme",
              },
            },
            "typecheck": {
              "cache": true,
              "command": "tsc --noEmit",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "metadata": {
                "description": "Runs type-checking for the project.",
                "help": {
                  "command": "pnpm exec tsc --help",
                  "example": {
                    "options": {
                      "noEmit": true,
                    },
                  },
                },
                "technologies": [
                  "typescript",
                ],
              },
              "options": {
                "cwd": "acme",
              },
            },
            "watch-deps": {
              "command": "pnpm exec nx watch --projects acme --includeDependentProjects -- pnpm exec nx build-deps acme",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/react/react-router-plugin React Router root project should create nodes by default 1`] = `
[
  [
    "react-router.config.js",
    {
      "projects": {
        ".": {
          "metadata": {},
          "root": ".",
          "targets": {
            "build": {
              "cache": true,
              "command": "react-router build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "react-router",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
              "outputs": [
                "{workspaceRoot}/build/client",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "react-router dev",
              "options": {
                "cwd": ".",
              },
            },
            "typecheck": {
              "cache": true,
              "command": "tsc --noEmit",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "metadata": {
                "description": "Runs type-checking for the project.",
                "help": {
                  "command": "pnpm exec tsc --help",
                  "example": {
                    "options": {
                      "noEmit": true,
                    },
                  },
                },
                "technologies": [
                  "typescript",
                ],
              },
              "options": {
                "cwd": ".",
              },
            },
            "watch-deps": {
              "command": "pnpm exec nx watch --projects acme --includeDependentProjects -- pnpm exec nx build-deps acme",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;

exports[`@nx/react/react-router-plugin React Router root project should create nodes without start target if ssr is false 1`] = `
[
  [
    "react-router.config.js",
    {
      "projects": {
        ".": {
          "metadata": {},
          "root": ".",
          "targets": {
            "build": {
              "cache": true,
              "command": "react-router build",
              "dependsOn": [
                "^build",
              ],
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "react-router",
                  ],
                },
              ],
              "options": {
                "cwd": ".",
              },
              "outputs": [
                "{workspaceRoot}/build/client",
              ],
            },
            "build-deps": {
              "dependsOn": [
                "^build",
              ],
            },
            "dev": {
              "command": "react-router dev",
              "options": {
                "cwd": ".",
              },
            },
            "typecheck": {
              "cache": true,
              "command": "tsc --noEmit",
              "inputs": [
                "production",
                "^production",
                {
                  "externalDependencies": [
                    "typescript",
                  ],
                },
              ],
              "metadata": {
                "description": "Runs type-checking for the project.",
                "help": {
                  "command": "pnpm exec tsc --help",
                  "example": {
                    "options": {
                      "noEmit": true,
                    },
                  },
                },
                "technologies": [
                  "typescript",
                ],
              },
              "options": {
                "cwd": ".",
              },
            },
            "watch-deps": {
              "command": "pnpm exec nx watch --projects acme --includeDependentProjects -- pnpm exec nx build-deps acme",
              "dependsOn": [
                "build-deps",
              ],
            },
          },
        },
      },
    },
  ],
]
`;
